service/round.go

```diff
@@ -191,108 +145,75 @@ func (r *round) persistExecution(tree *merkle.Tree, treeCache *cache.Writer, num
-	} else {
-		var err error
-		r.execution.Members, r.execution.Statement, err = r.calcMembersAndStatement()
+	if r.Execution.Members == nil || r.Execution.Statement == nil {
+		logger.Debug("calculating members and statement")
+		members, statement, err := r.calcMembersAndStatement()
 		if err != nil {
-			return err
+			return fmt.Errorf("failed to calculate members and statement")
 		}
+		r.Execution.Members, r.Execution.Statement = members, statement
 		if err := r.saveState(); err != nil {
 			return err
 		}
 	}
 
-	var err error
-	r.execution.NumLeaves, r.execution.NIP, err = prover.GenerateProofRecovery(
+	numLeaves, nip, err := prover.GenerateProofRecovery(
 		ctx,
 		r.datadir,
-		hash.GenLabelHashFunc(state.Statement),
-		hash.GenMerkleHashFunc(state.Statement),
+		hash.GenLabelHashFunc(r.Execution.Statement),
+		hash.GenMerkleHashFunc(r.Execution.Statement),
 		end,
-		state.SecurityParam,
-		state.NumLeaves,
-		state.ParkedNodes,
+		r.Execution.SecurityParam,
+		r.Execution.NumLeaves,
+		r.Execution.ParkedNodes,
 		r.persistExecution,
 	)
 	if err != nil {
 		return err
 	}
+	r.Execution.NumLeaves, r.Execution.NIP = numLeaves, nip
 	if err := r.saveState(); err != nil {
 		return err
 	}
 
-	close(r.executionEndedChan)
+	logger.With().Info("finished round recovered execution", log.Field(zap.Duration("duration", time.Since(started))))
 
 	return nil
 }
 
-func (r *round) proof(wait bool) (*PoetProof, error) {
-	if wait {
-		<-r.executionEndedChan
-	} else {
-		select {
-		case <-r.executionEndedChan:
-		default:
-			select {
-			case <-r.executionStartedChan:
-				return nil, errors.New("round is executing")
-			default:
-				select {
-				case <-r.openedChan:
-					return nil, errors.New("round is open")
-				default:
-					return nil, errors.New("round wasn't open")
-				}
-			}
-		}
-	}
-
-	return &PoetProof{
-		N:         uint(r.execution.NumLeaves),
-		Statement: r.execution.Statement,
-		Proof:     r.execution.NIP,
-	}, nil
-}
-
-func (r *round) state() (*roundState, error) {
+// loadState recovers persisted state from disk.
+func (r *round) loadState() error {
 	filename := filepath.Join(r.datadir, roundStateFileBaseName)
-	s := &roundState{}
-
-	if err := load(filename, s); err != nil {
-		return nil, err
+	state := roundState{}
+	if err := load(filename, &state); err != nil {
+		return err
 	}
-	if r.execution.SecurityParam != s.Execution.SecurityParam {
-		return nil, errors.New("SecurityParam config mismatch")
+	if r.Execution.SecurityParam != state.Execution.SecurityParam {
+		return errors.New("SecurityParam config mismatch")
 	}
-	r.stateCache = s
+	r.roundState = state
 
-	return s, nil
+	return nil
 }
 
 func (r *round) saveState() error {
 	filename := filepath.Join(r.datadir, roundStateFileBaseName)
-	v := &roundState{
-		Opened:           r.opened,
-		ExecutionStarted: r.executionStarted,
-		Execution:        r.execution,
-	}
-	return persist(filename, v)
+	return persist(filename, &r.roundState)
```

Comment:
same here, I'd prefer the way it was before: create a `roundState` object that is used to persist the state, but is not part of the round itself.

Reply:
Changed. PTAL